### PR TITLE
Make re-building __CPROVER_initialize safe

### PR DESCRIPTION
**Pull request by tautschnig**
We auto-generate `__CPROVER_initialize`, but also have/had several places that tweak the generated function. Such tweaking could be lost when rebuilding `__CPROVER_initialize`. The commits in this PR make sure that rebuilding `__CPROVER_initialize` is safe by making the prior tweaks permanent through updates to the symbol table.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by peterschrammel — APPROVED** [review]

File: src/goto-instrument/contracts/memory_predicates.cpp [hunk 1]
```diff
@@ -188,21 +190,18 @@ void is_fresh_baset::add_declarations(const std::string &decl_string)
...
+    static_lifetime_init(
+      parent.get_symbol_table(),
+      parent.get_symbol_table().lookup_ref(INITIALIZE_FUNCTION).location);
+    goto_convert(
+      INITIALIZE_FUNCTION,
+      parent.get_symbol_table(),
+      parent.get_goto_functions(),
+      log.get_message_handler());
+    parent.get_goto_functions().update();
+  }
```
Comment (feliperodri) [inline src/goto-instrument/contracts/memory_predicates.cpp]: Nice! It'll remove unnecessary warnings from our proofs with contracts.

**Review by feliperodri — APPROVED** [review]
Approving changes in our contract module.

File: src/cpp/cpp_internal_additions.cpp [hunk 2]
```diff
@@ -44,7 +44,7 @@ std::string c2cpp(const std::string &s)
 
 void cpp_internal_additions(std::ostream &out)
 {
-  out << "# 1 \"<built-in-additions>\"" << '\n';
+  out << "#line 1 \"<built-in-additions>\"" << '\n';
```
Comment (chrisr-diffblue) [inline src/cpp/cpp_internal_additions.cpp:47]: Good catch!

File: src/goto-cc/linker_script_merge.cpp [hunk 3]
```diff
@@ -453,24 +446,46 @@ int linker_script_merget::ls_data2instructions(
...
                   << array_size << " bytes";
     array_loc.set_comment(array_comment.str());
-    array_expr.add_source_location()=array_loc;
+
+    namespacet ns(symbol_table);
+    const auto zi = zero_initializer(array_type, array_loc, ns);
+    CHECK_RETURN(zi.has_value());
+
+    // Add the array to the symbol table.
+    symbolt array_sym;
```
Comment (chrisr-diffblue) [inline src/goto-cc/linker_script_merge.cpp:462]: Mildly a shame there isn't a constructor/utility function for creating a suitably initialised `symbolt`

**Review by chrisr-diffblue — APPROVED** [review]
Looks good :-)

File: src/goto-cc/linker_script_merge.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-cc/linker_script_merge.cpp:462]: Thank you for calling this out! I'll try to address this in a separate PR.

File: src/goto-instrument/goto_instrument_parse_options.h [hunk 4]
```diff
@@ -87,7 +88,6 @@ Author: Daniel Kroening, kroening@kroening.com
   "(show-symbol-table)(show-points-to)(show-rw-set)" \
   OPT_TIMESTAMP \
   "(show-natural-loops)(show-lexical-loops)(accelerate)(havoc-loops)" \
-  "(string-abstraction)" \
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.h:90]: @tautschnig Why has this been removed?

Comment (tautschnig) [reply src/goto-instrument/goto_instrument_parse_options.h:90]: Oh, my apologies for the obscure change: this is handled by `OPT_CONFIG_LIBRARY`.

Comment (martin-cs) [reply src/goto-instrument/goto_instrument_parse_options.h:90]: That's OK.  Now you mention it I have a nasty feeling I might have done that so... 😐

**Pull request merged**